### PR TITLE
feat: halving reward

### PR DESCRIPTION
**Pull request by zhangsoledad**

### CODE REVIEW

File: spec/src/consensus.rs [hunk 1]
```diff
@@ -654,6 +646,14 @@ impl Consensus {
             })
             .expect("Can not find secp script")
     }
+
+    fn primary_epoch_reward_of_next_epoch(&self, epoch: &EpochExt) -> Capacity {
+        if (epoch.number() + 1) % self.primary_epoch_reward_halving_interval() != 0 {
+            epoch.primary_reward()
+        } else {
+            Capacity::shannons(epoch.primary_reward().as_u64() / 2)
```
Comment (keroro520) [inline spec/src/consensus.rs]:
```suggestion
            self.primary_epoch_reward(epoch.number() + 1)
```

File: verification/src/transaction_verifier.rs [hunk 2]
```diff
@@ -183,7 +183,9 @@ impl<'a> EmptyVerifier<'a> {
     }
 
     pub fn verify(&self) -> Result<(), Error> {
-        if self.transaction.is_empty() {
+        if self.transaction.inputs().is_empty()
+            || (self.transaction.outputs().is_empty() && !self.transaction.is_cellbase())
+        {
```
Comment (keroro520) [inline verification/src/transaction_verifier.rs]: After this change, `TransactionView::is_empty` is useless. We can remove this function now.

File: util/types/src/core/extras.rs [hunk 3]
```diff
@@ -144,10 +149,17 @@ impl EpochExt {
         self.start_number = start_number;
     }
 
+    // Set length but keep the primary epoch rewards unchanged.
```
Comment (doitian) [inline util/types/src/core/extras.rs]: This comment is outdated and should be deleted.

File: chain/src/tests/basic.rs [hunk 4]
```diff
@@ -216,7 +216,7 @@ fn test_transaction_conflict_in_same_block() {
     let mut chain = MockChain::new(parent.clone(), shared.consensus());
     chain.gen_empty_block(&mock_store);
 
-    let last_cell_base = &chain.tip().transactions()[0];
+    let last_cell_base = &shared.consensus().genesis_block().transactions()[1];;
```
Comment (u2) [inline chain/src/tests/basic.rs]: why change the tip to genesis?

File: verification/src/contextual_block_verifier.rs [hunk 5]
```diff
@@ -248,22 +249,43 @@ impl<'a, 'b, CS: ChainStore<'a>> RewardVerifier<'a, 'b, CS> {
...
+                .transaction
+                .outputs()
+                .get(0)
+                .expect("cellbase should have output")
+                .lock()
+                != target_lock
+            {
+                return Err((CellbaseError::InvalidRewardTarget).into());
+            }
         }
```
Comment (u2) [inline verification/src/contextual_block_verifier.rs]: Add one more assertion that if it's `insufficient_reward_to_create_cell` the outputs must be empty.

File: chain/src/tests/basic.rs (hunk 4 above)
Comment (zhangsoledad) [reply chain/src/tests/basic.rs]: 1~11 block's cellbase has empty output,  can't be used to generate a new transaction. this broken many test.
I create some cell in genesis on mockchain to replace it.

Comment (u2) [reply chain/src/tests/basic.rs]: ok

**Comment by zhangsoledad** [comment]
bors r=u2,doitian,keroro520

**Comment by zhangsoledad** [comment]
bors r=u2,doitian,keroro520

**Comment by zhangsoledad** [comment]
bors retry

**Pull request merged**